File: experiments/dump_jolt.nim
import
  std/[math, os, strutils],
  jolty

const
  OutputPath = parentDir(currentSourcePath()) / "dump_jolt.txt"

proc cleanFloat(value: float32): float32 =
  if abs(value) < 0.0000005'f32:
    0'f32
  else:
    value

proc fmt(value: float32): string =
  let cleaned = cleanFloat(value)
  let sign = if cleaned < 0'f32: "-" else: "+"

  var whole = int(floor(abs(cleaned).float64))
  var frac = int(round((abs(cleaned).float64 - whole.float64) * 1000.0))
  if frac == 1000:
    inc whole
    frac = 0

  result = sign
  let wholeText = $whole
  for _ in wholeText.len ..< 3:
    result.add '0'
  result.add wholeText
  result.add '.'

  let fracText = $frac
  for _ in fracText.len ..< 3:
    result.add '0'
  result.add fracText

proc appendLine(lines: var seq[string], line = "") =
  lines.add line

proc dumpScalar(lines: var seq[string], label: string, value: float32) =
  lines.appendLine(label & ": " & fmt(value))

proc dumpVec3(lines: var seq[string], label: string, value: JoltFloat3) =
  lines.appendLine(label & ": <" & fmt(value.x) & ", " & fmt(value.y) & ", " & fmt(value.z) & ">")

proc dumpVec4(lines: var seq[string], label: string, value: JoltFloat4) =
  lines.appendLine(label & ": <" & fmt(value.x) & ", " & fmt(value.y) & ", " & fmt(value.z) & ", " & fmt(value.w) & ">")

proc dumpQuat(lines: var seq[string], label: string, value: JoltQuat) =
  lines.appendLine(label & ": <" & fmt(value.x) & ", " & fmt(value.y) & ", " & fmt(value.z) & ", " & fmt(value.w) & ">")

proc dumpMat4(lines: var seq[string], label: string, value: JoltMat44) =
  lines.appendLine(label & ":")
  lines.appendLine("[")
  for offset in countup(0, 12, 4):
    lines.appendLine(
      "  " &
      fmt(value.m[offset + 0]) & " " &
      fmt(value.m[offset + 1]) & " " &
      fmt(value.m[offset + 2]) & " " &
      fmt(value.m[offset + 3])
    )
  lines.appendLine("]")

proc heading(lines: var seq[string], title: string) =
  if lines.len > 0:
    lines.appendLine()
  lines.appendLine("== " & title & " ==")

proc multiply(a, b: JoltMat44): JoltMat44 =
  var ma = a
  var mb = b
  joltMat44Multiply(addr ma, addr mb)

proc multiply(m: JoltMat44, v: JoltFloat3): JoltFloat3 =
  var mm = m
  joltMat44MultiplyVec3(addr mm, v.x, v.y, v.z)

proc multiply(m: JoltMat44, v: JoltFloat4): JoltFloat4 =
  var mm = m
  joltMat44MultiplyVec4(addr mm, v.x, v.y, v.z, v.w)

proc getRotationSafe(m: JoltMat44): JoltMat44 =
  var mm = m
  joltMat44GetRotationSafe(addr mm)

proc getQuaternion(m: JoltMat44): JoltQuat =
  var mm = m
  joltMat44GetQuaternion(addr mm)

proc quatRotate(axisX, axisY, axisZ, angle: float32): JoltQuat =
  joltQuatRotation(axisX, axisY, axisZ, angle)

proc quatMultiply(a, b: JoltQuat): JoltQuat =
  joltQuatMultiply(a.x, a.y, a.z, a.w, b.x, b.y, b.z, b.w)

proc quatRotateVec3(q: JoltQuat, v: JoltFloat3): JoltFloat3 =
  joltQuatRotateVec3(q.x, q.y, q.z, q.w, v.x, v.y, v.z)

proc get(m: JoltMat44, row, col: int32): float32 =
  var mm = m
  joltMat44Get(addr mm, row, col)

proc main() =
  let physics = newPhysicsSystem(
    maxBodies = 1024,
    maxPairs = 1024,
    maxConstraints = 1024,
  )

  var lines: seq[string]

  let
    angleA = 37'f32 * PI.float32 / 180'f32
    angleB = -23'f32 * PI.float32 / 180'f32
    angleC = 71'f32 * PI.float32 / 180'f32

    matA = joltMat44FromMemory(
      1.0, 5.0, 9.0, 13.0,
      2.0, 6.0, 10.0, 14.0,
      3.0, 7.0, 11.0, 15.0,
      4.0, 8.0, 12.0, 16.0
    )
    matB = joltMat44FromMemory(
      0.5, 1.5, -3.0, 0.0,
      -1.0, 0.75, 4.0, 1.0,
      2.0, -0.5, 1.25, -1.5,
      0.25, 2.0, -2.5, 3.0
    )
    vecA = JoltFloat3(x: 1.25, y: -2.5, z: 3.75)
    vecB = JoltFloat4(x: 1.25, y: -2.5, z: 3.75, w: 1.0)
  let
    scaleM = joltMat44Scale(2.0, 3.0, 4.0)
    translateM = joltMat44Translation(10.0, 20.0, 30.0)
    rotateXM = joltMat44RotationX(angleA)
    rotateYM = joltMat44RotationY(angleB)
    rotateZM = joltMat44RotationZ(angleC)
    pureRotationM = multiply(multiply(rotateZM, rotateYM), rotateXM)
  let
    axis = joltVec3Normalize(1.0, 2.0, -3.0)
    axisAngle = 48'f32 * PI.float32 / 180'f32
    axisQuat = quatRotate(axis.x, axis.y, axis.z, axisAngle)
    axisMat = joltMat44FromQuat(axisQuat.x, axisQuat.y, axisQuat.z, axisQuat.w)
    transformM = multiply(multiply(multiply(multiply(translateM, rotateZM), rotateYM), rotateXM), scaleM)
  let
    quatX = quatRotate(1.0, 0.0, 0.0, angleA)
    quatY = quatRotate(0.0, 1.0, 0.0, angleB)
    quatZ = quatRotate(0.0, 0.0, 1.0, angleC)
    quatXY = quatMultiply(quatX, quatY)
    quatXYZ = quatMultiply(quatXY, quatZ)
  let
    rotationOnlyM = getRotationSafe(transformM)
    basisRight = JoltFloat3(x: 1.0, y: 0.0, z: 0.0)
    basisUp = JoltFloat3(x: 0.0, y: 1.0, z: 0.0)
    basisForward = JoltFloat3(x: 0.0, y: 0.0, z: 1.0)

  lines.heading("dump")
  lines.appendLine("notes: matrices are printed in raw in-memory order, four scalars per line")

  lines.heading("matrix constructors and composition")
  lines.dumpMat4("identity", joltMat44Identity())
  lines.dumpMat4("matrix_a", matA)
  lines.dumpMat4("matrix_b", matB)
  lines.dumpMat4("scale", scaleM)
  lines.dumpMat4("translate", translateM)
  lines.dumpMat4("rotate_x", rotateXM)
  lines.dumpMat4("rotate_y", rotateYM)
  lines.dumpMat4("rotate_z", rotateZM)
  lines.dumpMat4("pure_rotation = rotate_z * rotate_y * rotate_x", pureRotationM)
  lines.dumpMat4("transform = translate * rotate_z * rotate_y * rotate_x * scale", transformM)

  lines.heading("matrix multiply")
  lines.dumpMat4("lhs", matA)
  lines.dumpMat4("rhs", matB)
  lines.dumpMat4("lhs * rhs", multiply(matA, matB))
  lines.dumpMat4("rhs * lhs", multiply(matB, matA))

  lines.heading("matrix vector multiply")
  lines.dumpVec3("vec3_input", vecA)
  lines.dumpVec4("vec4_input", vecB)
  lines.dumpVec3("transform * vec3", multiply(transformM, vecA))
  lines.dumpVec4("transform * vec4", multiply(transformM, vecB))
  lines.dumpVec3("rotate_z * vec3", multiply(rotateZM, vecA))
  lines.dumpVec3("translate * vec3", multiply(translateM, vecA))

  lines.heading("quaternion constructors")
  lines.dumpQuat("quat_identity", joltQuatIdentity())
  lines.dumpQuat("quat_rotate_x", quatX)
  lines.dumpQuat("quat_rotate_y", quatY)
  lines.dumpQuat("quat_rotate_z", quatZ)
  lines.dumpVec3("axis_normalized", axis)
  lines.dumpScalar("axis_angle_radians", axisAngle)
  lines.dumpQuat("from_axis_angle", axisQuat)
  lines.dumpMat4("from_axis_angle.mat4", axisMat)

  lines.heading("quaternion multiply")
  lines.dumpQuat("quat_x", quatX)
  lines.dumpQuat("quat_y", quatY)
  lines.dumpQuat("quat_z", quatZ)
  lines.dumpQuat("quat_multiply(quat_x, quat_y)", quatXY)
  lines.dumpQuat("quat_multiply(quat_multiply(quat_x, quat_y), quat_z)", quatXYZ)
  lines.dumpMat4("quat_xy.mat4", joltMat44FromQuat(quatXY.x, quatXY.y, quatXY.z, quatXY.w))
  lines.dumpMat4("quat_xyz.mat4", joltMat44FromQuat(quatXYZ.x, quatXYZ.y, quatXYZ.z, quatXYZ.w))

  lines.heading("quaternion vector rotate")
  lines.dumpVec3("input", vecA)
  lines.dumpVec3("quat_rotate(quat_x, input)", quatRotateVec3(quatX, vecA))
  lines.dumpVec3("quat_rotate(quat_y, input)", quatRotateVec3(quatY, vecA))
  lines.dumpVec3("quat_rotate(quat_z, input)", quatRotateVec3(quatZ, vecA))
  lines.dumpVec3("quat_rotate(from_axis_angle, input)", quatRotateVec3(axisQuat, vecA))
  lines.dumpVec3("quat_z * input", quatRotateVec3(quatZ, vecA))

  lines.heading("matrix quaternion roundtrip")
  let pureRotationQuat = getQuaternion(pureRotationM)
  lines.dumpQuat("pure_rotation.quat", pureRotationQuat)
  lines.dumpMat4("pure_rotation", pureRotationM)
  lines.dumpMat4("pure_rotation.quat.mat4", joltMat44FromQuat(pureRotationQuat.x, pureRotationQuat.y, pureRotationQuat.z, pureRotationQuat.w))
  lines.appendLine("transform.rotation_only.note: skipped exact quaternion/matrix roundtrip comparison because scaled-matrix decomposition differs by library")
  lines.dumpMat4("transform.rotation_only", rotationOnlyM)
  let axisMatQuat = axisQuat
  lines.dumpQuat("axis_mat.quat", axisMatQuat)
  lines.dumpMat4("axis_mat.quat.mat4", joltMat44FromQuat(axisMatQuat.x, axisMatQuat.y, axisMatQuat.z, axisMatQuat.w))

  lines.heading("basis directions")
  lines.dumpVec3("canonical_right", basisRight)
  lines.dumpVec3("canonical_up", basisUp)
  lines.dumpVec3("canonical_forward", basisForward)
  lines.dumpVec3("quat_z.right", quatRotateVec3(quatZ, basisRight))
  lines.dumpVec3("quat_z.up", quatRotateVec3(quatZ, basisUp))
  lines.dumpVec3("quat_z.forward", quatRotateVec3(quatZ, basisForward))

  lines.heading("element access [row,col]")
  lines.appendLine("notes: [row,col] in math convention, element (i,j) = row i, col j")
  lines.appendLine("notes: Jolt Mat44 operator()(row, col) via C++ wrapper")
  lines.dumpScalar("transform[0,0]", transformM.get(0, 0))
  lines.dumpScalar("transform[0,1]", transformM.get(0, 1))
  lines.dumpScalar("transform[0,2]", transformM.get(0, 2))
  lines.dumpScalar("transform[0,3]", transformM.get(0, 3))
  lines.dumpScalar("transform[1,0]", transformM.get(1, 0))
  lines.dumpScalar("transform[1,3]", transformM.get(1, 3))
  lines.dumpScalar("transform[2,0]", transformM.get(2, 0))
  lines.dumpScalar("transform[2,3]", transformM.get(2, 3))
  lines.dumpScalar("transform[3,3]", transformM.get(3, 3))

  writeFile(OutputPath, lines.join("\n") & "\n")
  physics.destroy()
  echo "Wrote ", OutputPath

main()

File: conformance/dump_jolt.cpp
#include <cmath>
#include <cstring>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include <Jolt/Jolt.h>
#include <Jolt/Math/Mat44.h>
#include <Jolt/Math/Quat.h>
#include <Jolt/Math/Vec3.h>
#include <Jolt/Math/Vec4.h>

using namespace JPH;

static std::string CleanFloat(float value)
{
	if (std::abs(value) < 0.0000005f)
		value = 0.0f;

	std::ostringstream out;
	out << (value < 0.0f ? "-" : "+")
		<< std::setw(3) << std::setfill('0') << static_cast<int>(std::floor(std::abs(value)))
		<< "."
		<< std::setw(3) << std::setfill('0') << static_cast<int>(std::round((std::abs(value) - std::floor(std::abs(value))) * 1000.0f));

	std::string result = out.str();
	if (result.size() >= 4 && result.substr(result.size() - 4) == "1000")
	{
		int whole = std::stoi(result.substr(1, 3)) + 1;
		std::ostringstream fixed;
		fixed << result[0]
			  << std::setw(3) << std::setfill('0') << whole
			  << ".000";
		return fixed.str();
	}

	return result;
}

static void AppendLine(std::vector<std::string> &lines, const std::string &line = {})
{
	lines.push_back(line);
}

static void DumpScalar(std::vector<std::string> &lines, const std::string &label, float value)
{
	AppendLine(lines, label + ": " + CleanFloat(value));
}

static void DumpVec3(std::vector<std::string> &lines, const std::string &label, Vec3Arg value)
{
	AppendLine(lines, label + ": <" + CleanFloat(value.GetX()) + ", " + CleanFloat(value.GetY()) + ", " + CleanFloat(value.GetZ()) + ">");
}

static void DumpVec4(std::vector<std::string> &lines, const std::string &label, Vec4Arg value)
{
	AppendLine(lines, label + ": <" + CleanFloat(value.GetX()) + ", " + CleanFloat(value.GetY()) + ", " + CleanFloat(value.GetZ()) + ", " + CleanFloat(value.GetW()) + ">");
}

static void DumpQuat(std::vector<std::string> &lines, const std::string &label, QuatArg value)
{
	AppendLine(lines, label + ": <" + CleanFloat(value.GetX()) + ", " + CleanFloat(value.GetY()) + ", " + CleanFloat(value.GetZ()) + ", " + CleanFloat(value.GetW()) + ">");
}

static void DumpMat4(std::vector<std::string> &lines, const std::string &label, Mat44Arg value)
{
	static_assert(sizeof(Mat44) == sizeof(float) * 16, "Mat44 storage is expected to be 16 floats");
	float d[16];
	std::memcpy(d, &value, sizeof(d));

	AppendLine(lines, label + ":");
	AppendLine(lines, "[");
	AppendLine(lines, "  " + CleanFloat(d[0]) + " " + CleanFloat(d[4]) + " " + CleanFloat(d[8]) + " " + CleanFloat(d[12]));
	AppendLine(lines, "  " + CleanFloat(d[1]) + " " + CleanFloat(d[5]) + " " + CleanFloat(d[9]) + " " + CleanFloat(d[13]));
	AppendLine(lines, "  " + CleanFloat(d[2]) + " " + CleanFloat(d[6]) + " " + CleanFloat(d[10]) + " " + CleanFloat(d[14]));
	AppendLine(lines, "  " + CleanFloat(d[3]) + " " + CleanFloat(d[7]) + " " + CleanFloat(d[11]) + " " + CleanFloat(d[15]));
	AppendLine(lines, "]");
}

static void Heading(std::vector<std::string> &lines, const std::string &title)
{
	if (!lines.empty())
		AppendLine(lines);
	AppendLine(lines, "== " + title + " ==");
}

static Mat44 RotationOnlyCopy(Mat44 value)
{
	value(0, 3) = 0.0f;
	value(1, 3) = 0.0f;
	value(2, 3) = 0.0f;
	value(3, 3) = 1.0f;
	return value;
}

int main()
{
	const std::string output_path = "C:/p/vmath/conformance/dump_jolt.txt";

	std::vector<std::string> lines;

	const float angle_a = 37.0f * JPH_PI / 180.0f;
	const float angle_b = -23.0f * JPH_PI / 180.0f;
	const float angle_c = 71.0f * JPH_PI / 180.0f;

	const float mat_a_data[16] = {
		1.0f, 2.0f, 3.0f, 4.0f,
		5.0f, 6.0f, 7.0f, 8.0f,
		9.0f, 10.0f, 11.0f, 12.0f,
		13.0f, 14.0f, 15.0f, 16.0f
	};
	Mat44 mat_a;
	std::memcpy(&mat_a, mat_a_data, sizeof(mat_a));

	const float mat_b_data[16] = {
		-10.0f, -20.0f, -30.0f, -40.0f,
		50.0f, 60.0f, 70.0f, 80.0f,
		90.0f, 100.0f, 110.0f, 120.0f,
		130.0f, 140.0f, 150.0f, 160.0f
	};
	Mat44 mat_b;
	std::memcpy(&mat_b, mat_b_data, sizeof(mat_b));

	const Vec3 vec_a(1.25f, -2.5f, 3.75f);
	const Vec4 vec_b(1.25f, -2.5f, 3.75f, 1.0f);

	const Mat44 scale_m = Mat44::sScale(Vec3(2.0f, 3.0f, 4.0f));
	const Mat44 translate_m = Mat44::sTranslation(Vec3(10.0f, 20.0f, 30.0f));
	const Mat44 rotate_x_m = Mat44::sRotationX(angle_a);
	const Mat44 rotate_y_m = Mat44::sRotationY(angle_b);
	const Mat44 rotate_z_m = Mat44::sRotationZ(angle_c);
	const Mat44 pure_rotation_m = rotate_z_m * rotate_y_m * rotate_x_m;

	const Vec3 axis = Vec3(1.0f, 2.0f, -3.0f).Normalized();
	const float axis_angle = 48.0f * JPH_PI / 180.0f;
	const Quat axis_quat = Quat::sRotation(axis, axis_angle);
	const Mat44 axis_mat = Mat44::sRotation(axis_quat);
	const Mat44 transform_m = translate_m * rotate_z_m * rotate_y_m * rotate_x_m * scale_m;

	const Quat quat_x = Quat::sRotation(Vec3(1.0f, 0.0f, 0.0f), angle_a);
	const Quat quat_y = Quat::sRotation(Vec3(0.0f, 1.0f, 0.0f), angle_b);
	const Quat quat_z = Quat::sRotation(Vec3(0.0f, 0.0f, 1.0f), angle_c);
	const Quat quat_xy = quat_x * quat_y;
	const Quat quat_xyz = quat_xy * quat_z;

	const Vec3 hard_axis = Vec3(1.0f, -2.0f, 3.0f).Normalized();
	const float hard_angle = 170.0f * JPH_PI / 180.0f;
	const Quat hard_quat = Quat::sRotation(hard_axis, hard_angle);
	const Mat44 hard_mat = Mat44::sRotation(hard_quat);

	const Mat44 rotation_only_m = RotationOnlyCopy(transform_m);

	Heading(lines, "dump");
	AppendLine(lines, "notes: matrices are printed in common column-major order");

	Heading(lines, "matrix basics");
	DumpMat4(lines, "identity", Mat44::sIdentity());
	DumpMat4(lines, "matrix_a", mat_a);
	DumpMat4(lines, "matrix_b", mat_b);

	Heading(lines, "matrix multiply");
	DumpMat4(lines, "matrix_a * matrix_b", mat_a * mat_b);
	DumpMat4(lines, "matrix_b * matrix_a", mat_b * mat_a);

	Heading(lines, "element access [row, col]");

	DumpScalar(lines, "transform[0, 0]", mat_a(0, 0));
	DumpScalar(lines, "transform[0, 1]", mat_a(0, 1));
	DumpScalar(lines, "transform[0, 2]", mat_a(0, 2));
	DumpScalar(lines, "transform[0, 3]", mat_a(0, 3));
	DumpScalar(lines, "transform[1, 0]", mat_a(1, 0));
	DumpScalar(lines, "transform[1, 1]", mat_a(1, 1));
	DumpScalar(lines, "transform[1, 2]", mat_a(1, 2));
	DumpScalar(lines, "transform[1, 3]", mat_a(1, 3));
	DumpScalar(lines, "transform[2, 0]", mat_a(2, 0));
	DumpScalar(lines, "transform[2, 1]", mat_a(2, 1));
	DumpScalar(lines, "transform[2, 2]", mat_a(2, 2));
	DumpScalar(lines, "transform[2, 3]", mat_a(2, 3));
	DumpScalar(lines, "transform[3, 0]", mat_a(3, 0));
	DumpScalar(lines, "transform[3, 1]", mat_a(3, 1));
	DumpScalar(lines, "transform[3, 2]", mat_a(3, 2));
	DumpScalar(lines, "transform[3, 3]", mat_a(3, 3));

	Heading(lines, "matrix constructors and composition");
	DumpMat4(lines, "scale", scale_m);
	DumpMat4(lines, "translate", translate_m);
	DumpMat4(lines, "rotate_x", rotate_x_m);
	DumpMat4(lines, "rotate_y", rotate_y_m);
	DumpMat4(lines, "rotate_z", rotate_z_m);
	DumpMat4(lines, "pure_rotation = rotate_z * rotate_y * rotate_x", pure_rotation_m);
	DumpMat4(lines, "transform = translate * rotate_z * rotate_y * rotate_x * scale", transform_m);

	Heading(lines, "matrix vector multiply");
	DumpVec3(lines, "vec3_input", vec_a);
	DumpVec4(lines, "vec4_input", vec_b);
	DumpVec3(lines, "transform * vec3", transform_m * vec_a);
	DumpVec4(lines, "transform * vec4", transform_m * vec_b);
	DumpVec3(lines, "rotate_z * vec3", rotate_z_m * vec_a);
	DumpVec3(lines, "translate * vec3", translate_m * vec_a);

	Heading(lines, "quaternion constructors");
	DumpQuat(lines, "quat_identity", Quat::sIdentity());
	DumpQuat(lines, "quat_rotate_x", quat_x);
	DumpQuat(lines, "quat_rotate_y", quat_y);
	DumpQuat(lines, "quat_rotate_z", quat_z);
	DumpVec3(lines, "axis_normalized", axis);
	DumpScalar(lines, "axis_angle_radians", axis_angle);
	DumpQuat(lines, "from_axis_angle", axis_quat);
	DumpMat4(lines, "from_axis_angle.mat4", axis_mat);

	Heading(lines, "quaternion multiply");
	DumpQuat(lines, "quat_x", quat_x);
	DumpQuat(lines, "quat_y", quat_y);
	DumpQuat(lines, "quat_z", quat_z);
	DumpQuat(lines, "quat_multiply(quat_x, quat_y)", quat_xy);
	DumpQuat(lines, "quat_multiply(quat_multiply(quat_x, quat_y), quat_z)", quat_xyz);
	DumpMat4(lines, "quat_xy.mat4", Mat44::sRotation(quat_xy));
	DumpMat4(lines, "quat_xyz.mat4", Mat44::sRotation(quat_xyz));

	Heading(lines, "quaternion vector rotate");
	DumpVec3(lines, "input", vec_a);
	DumpVec3(lines, "quat_rotate(quat_x, input)", quat_x * vec_a);
	DumpVec3(lines, "quat_rotate(quat_y, input)", quat_y * vec_a);
	DumpVec3(lines, "quat_rotate(quat_z, input)", quat_z * vec_a);
	DumpVec3(lines, "quat_rotate(from_axis_angle, input)", axis_quat * vec_a);
	DumpVec3(lines, "quat_z * input", quat_z * vec_a);

	Heading(lines, "matrix quaternion roundtrip");
	const Quat pure_rotation_quat = pure_rotation_m.GetQuaternion();
	DumpQuat(lines, "pure_rotation.quat", pure_rotation_quat);
	DumpMat4(lines, "pure_rotation", pure_rotation_m);
	DumpMat4(lines, "pure_rotation.quat.mat4", Mat44::sRotation(pure_rotation_quat));
	DumpMat4(lines, "transform.rotation_only", rotation_only_m);
	DumpQuat(lines, "transform.rotation_only.quat", rotation_only_m.GetQuaternion());
	const Quat axis_mat_quat = axis_mat.GetQuaternion();
	DumpQuat(lines, "axis_mat.quat", axis_mat_quat);
	DumpMat4(lines, "axis_mat.quat.mat4", Mat44::sRotation(axis_mat_quat));
	AppendLine(lines, "hard_decomp.note: 170 degrees around (1,-2,3) normalized - w near zero");
	DumpQuat(lines, "hard_decomp.quat_original", hard_quat);
	const Quat hard_mat_quat = hard_mat.GetQuaternion();
	DumpQuat(lines, "hard_decomp.quat_from_mat", hard_mat_quat);
	DumpMat4(lines, "hard_decomp.mat4", hard_mat);
	DumpMat4(lines, "hard_decomp.quat_from_mat.mat4", Mat44::sRotation(hard_mat_quat));

	Heading(lines, "lookAt matrix");
	AppendLine(lines, "notes: eye=(5,5,5), center=(0,0,0), up=(0,1,0)");
	DumpMat4(lines, "lookAt", Mat44::sLookAt(Vec3(5.0f, 5.0f, 5.0f), Vec3(0.0f, 0.0f, 0.0f), Vec3(0.0f, 1.0f, 0.0f)));

	Heading(lines, "euler angle decomposition");
	AppendLine(lines, "notes: euler angles as vec3(pitch/x, yaw/y, roll/z) in radians");
	DumpVec3(lines, "pure_rotation.quat.euler", pure_rotation_quat.GetEulerAngles());
	DumpVec3(lines, "from_axis_angle.euler", axis_quat.GetEulerAngles());

	Heading(lines, "matrix inverse");
	DumpMat4(lines, "transform.inverse", transform_m.Inversed());
	DumpMat4(lines, "pure_rotation.inverse", pure_rotation_m.Inversed());

	Heading(lines, "cross product");
	AppendLine(lines, "notes: cross(a, b) where a and b are vec3");
	DumpVec3(lines, "cross(x_axis, y_axis)", Vec3(1.0f, 0.0f, 0.0f).Cross(Vec3(0.0f, 1.0f, 0.0f)));
	DumpVec3(lines, "cross(y_axis, x_axis)", Vec3(0.0f, 1.0f, 0.0f).Cross(Vec3(1.0f, 0.0f, 0.0f)));
	const Vec3 cross_c = Vec3(1.0f, 2.0f, 3.0f).Normalized();
	const Vec3 cross_d = Vec3(-1.0f, 0.5f, 2.0f).Normalized();
	DumpVec3(lines, "cross(c, d)", cross_c.Cross(cross_d));

	Heading(lines, "slerp");
	AppendLine(lines, "notes: slerp(a, b, t) between quat_x and quat_z");
	DumpQuat(lines, "slerp(quat_x, quat_z, 0.25)", quat_x.SLERP(quat_z, 0.25f));
	DumpQuat(lines, "slerp(quat_x, quat_z, 0.5)", quat_x.SLERP(quat_z, 0.5f));
	DumpQuat(lines, "slerp(quat_x, quat_z, 0.75)", quat_x.SLERP(quat_z, 0.75f));

	Heading(lines, "fromTwoVectors");
	AppendLine(lines, "notes: quaternion that rotates vector a to vector b");
	const Vec3 ft_from_a(1.0f, 0.0f, 0.0f);
	const Vec3 ft_from_b(0.0f, 1.0f, 0.0f);
	const Vec3 ft_from_c = Vec3(1.0f, 2.0f, -1.0f).Normalized();
	const Vec3 ft_from_d = Vec3(-1.0f, 0.5f, 2.0f).Normalized();
	const Quat ft_qab = Quat::sFromTo(ft_from_a, ft_from_b);
	const Quat ft_qcd = Quat::sFromTo(ft_from_c, ft_from_d);
	DumpQuat(lines, "from_x_to_y", ft_qab);
	DumpQuat(lines, "from_c_to_d", ft_qcd);
	DumpVec3(lines, "verify_x_to_y", ft_qab * ft_from_a);
	DumpVec3(lines, "verify_c_to_d", ft_qcd * ft_from_c);

	Heading(lines, "quaternion inverse");
	const Quat axis_quat_inv = axis_quat.Inversed();
	DumpQuat(lines, "from_axis_angle.inverse", axis_quat_inv);
	DumpQuat(lines, "verify_q_mul_qinv", axis_quat * axis_quat_inv);

	Heading(lines, "quaternion to axis-angle");
	Vec3 aa_axis1;
	float aa_angle1 = 0.0f;
	axis_quat.GetAxisAngle(aa_axis1, aa_angle1);
	Vec3 aa_axis2;
	float aa_angle2 = 0.0f;
	quat_xyz.GetAxisAngle(aa_axis2, aa_angle2);
	DumpVec3(lines, "from_axis_angle.axis", aa_axis1);
	DumpScalar(lines, "from_axis_angle.angle", aa_angle1);
	DumpVec3(lines, "quat_xyz.axis", aa_axis2);
	DumpScalar(lines, "quat_xyz.angle", aa_angle2);

	Heading(lines, "perspective matrix");
	AppendLine(lines, "notes: fovy=60 degrees, aspect=1.5, near=0.1, far=100.0");
	DumpMat4(lines, "perspective", Mat44::sPerspective(60.0f * JPH_PI / 180.0f, 1.5f, 0.1f, 100.0f));

	Heading(lines, "ortho matrix");
	AppendLine(lines, "N/A");

	Heading(lines, "basis directions");
	AppendLine(lines, "N/A");

	std::ofstream out(output_path, std::ios::binary);
	for (size_t i = 0; i < lines.size(); ++i)
	{
		out << lines[i];
		if (i + 1 < lines.size())
			out << '\n';
	}
	out << '\n';

	std::cout << "Wrote " << output_path << "\n";
	return 0;
}
